mediaengine.go

```diff
@@ -4,419 +4,280 @@ package webrtc
-			if err != nil {
-				return fmt.Errorf("%w: codec for payload type %d", errMediaEngineCodecNotFound, payloadType)
-			}
-
-			var codec *RTPCodec
-			switch {
-			case strings.EqualFold(payloadCodec.Name, PCMA):
-				codec = NewRTPPCMACodec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, PCMU):
-				codec = NewRTPPCMUCodec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, G722):
-				codec = NewRTPG722Codec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, Opus):
-				codec = NewRTPOpusCodec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, VP8):
-				codec = NewRTPVP8Codec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, VP9):
-				codec = NewRTPVP9Codec(payloadType, payloadCodec.ClockRate)
-			case strings.EqualFold(payloadCodec.Name, H264):
-				codec = NewRTPH264Codec(payloadType, payloadCodec.ClockRate)
-			default:
-				// ignoring other codecs
-				continue
-			}
-
-			codec.SDPFmtpLine = payloadCodec.Fmtp
-			m.RegisterCodec(codec)
-		}
-	}
-	return nil
-}
-
-// GetCodecsByName returns all codecs by name that are supported by m.
-// The returned codecs should not be modified.
-func (m *MediaEngine) GetCodecsByName(codecName string) []*RTPCodec {
-	var codecs []*RTPCodec
-	for _, codec := range m.codecs {
-		if strings.EqualFold(codec.Name, codecName) {
-			codecs = append(codecs, codec)
-		}
-	}
-	return codecs
-}
-
-func (m *MediaEngine) getCodec(payloadType uint8) (*RTPCodec, error) {
-	for _, codec := range m.codecs {
-		if codec.PayloadType == payloadType {
-			return codec, nil
+func (m *MediaEngine) RegisterDefaultCodecs() error {
+	// Default Pion Audio Codecs
+	for _, codec := range []RTPCodecParameters{
+		{
+			RTPCodecCapability: RTPCodecCapability{mimeTypeOpus, 48000, 2, "minptime=10;useinbandfec=1", []RTCPFeedback{{"transport-cc", ""}}},
+			PayloadType:        111,
+		},
+		{
+			RTPCodecCapability: RTPCodecCapability{mimeTypeG722, 8000, 0, "", nil},
+			PayloadType:        9,
+		},
+		{
+			RTPCodecCapability: RTPCodecCapability{mimeTypePCMU, 8000, 0, "", nil},
+			PayloadType:        0,
+		},
+		{
+			RTPCodecCapability: RTPCodecCapability{mimeTypePCMA, 8000, 0, "", nil},
+			PayloadType:        8,
+		},
+	} {
+		if err := m.RegisterCodec(codec, RTPCodecTypeAudio); err != nil {
+			return err
 		}
 	}
-	return nil, ErrCodecNotFound
-}
 
-func (m *MediaEngine) getCodecSDP(sdpCodec sdp.Codec) (*RTPCodec, error) {
-	for _, codec := range m.codecs {
-		if strings.EqualFold(codec.Name, sdpCodec.Name) &&
-			codec.ClockRate == sdpCodec.ClockRate &&
-			(sdpCodec.EncodingParameters == "" ||
-				strconv.Itoa(int(codec.Channels)) == sdpCodec.EncodingParameters) &&
-			codec.SDPFmtpLine == sdpCodec.Fmtp { // pion/webrtc#43
-			return codec, nil
+	// Default Pion Audio Header Extensions
+	for _, extension := range []string{
```

Comment:
I think we should remove the following headers from default:

- "http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01"
- "urn:ietf:params:rtp-hdrext:sdes:mid"
- "urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id"
- "urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id"

If pion users do not implement the correct behavior in rtp packets for above headers, it may lead to errors that are hard to debug, like twcc will drop the bitrate to zero it's not implemented.

Reply:
:+1: nice, my mistake! Will pull those out.

Would it be ok to leave mid/rtp-stream-id/repaired-stream-id? Right now our Simulcast examples does REMB + those headers and seems ok. It would just be nice if users didn't have to touch headers when doing Simulcast.

Reply:
Simulcast will require the sdes:mid and rtp-stream-id, the problem with those, is that if they are always negotiated and pion is the media producer if users dont add correct sdes:mid in the header they may encounter cryptic erros in Chrome at least.

mid/rtp-stream-id/repaired-stream-id is for simulcast rtx (rrid) as far as I know, I think this should be optional also. Or not sure how rtx will be handled by Pion.

Reply:
Pulled everything transport-cc out!

I confirmed and mid/rtp-stream-id doesn't effect anything, so leaving in for now!